File: Arrays/493-Reverse-Pairs/JavaScript/493-Reverse-Pairs-Optimal.js
/**
 * Problem: 493. Reverse Pairs
 * Approach: Optimal - Merge Sort + Two Pointers
 *
 * Count pairs where:
 *     nums[i] > 2 * nums[j]
 *     i < j
 *
 * Time Complexity: O(n log n)
 * Space Complexity: O(n)
 */

/**
 * @param {number[]} nums
 * @return {number}
 */
var reversePairs = function(nums) {

    function countPairs(low, mid, high) {
        let count = 0;
        let right = mid + 1;

        for (let i = low; i <= mid; i++) {
            while (
                right <= high &&
                nums[i] > 2 * nums[right]
            ) {
                right++;
            }

            count += right - (mid + 1);
        }

        return count;
    }

    function merge(low, mid, high) {
        const temp = [];

        let left = low;
        let right = mid + 1;

        while (left <= mid && right <= high) {
            if (nums[left] <= nums[right]) {
                temp.push(nums[left]);
                left++;
            } else {
                temp.push(nums[right]);
                right++;
            }
        }

        while (left <= mid) {
            temp.push(nums[left]);
            left++;
        }

        while (right <= high) {
            temp.push(nums[right]);
            right++;
        }

        for (let i = 0; i < temp.length; i++) {
            nums[low + i] = temp[i];
        }
    }

    function mergeSort(low, high) {
        if (low >= high) {
            return 0;
        }

        const mid = Math.floor((low + high) / 2);

        let count = 0;

        count += mergeSort(low, mid);
        count += mergeSort(mid + 1, high);

        count += countPairs(low, mid, high);

        merge(low, mid, high);

        return count;
    }

    return mergeSort(0, nums.length - 1);
};

File: Arrays/493-Reverse-Pairs/Python/493-Reverse-Pairs-Optimal.py
"""
Problem: 493. Reverse Pairs
Approach: Optimal - Merge Sort + Two Pointers

First recursively sort both halves.

Before merging, count cross-half reverse pairs:
nums[i] > 2 * nums[j]

Since both halves are sorted, use a pointer on the
right half to count all valid pairs efficiently.

Time Complexity: O(n log n)
Space Complexity: O(n)
"""

class Solution:
    def reversePairs(self, nums: list[int]) -> int:

        def countPairs(low, mid, high):
            count = 0
            right = mid + 1

            for i in range(low, mid + 1):
                while right <= high and nums[i] > 2 * nums[right]:
                    right += 1

                count += right - (mid + 1)

            return count

        def merge(low, mid, high):
            temp = []

            left = low
            right = mid + 1

            while left <= mid and right <= high:
                if nums[left] <= nums[right]:
                    temp.append(nums[left])
                    left += 1
                else:
                    temp.append(nums[right])
                    right += 1

            while left <= mid:
                temp.append(nums[left])
                left += 1

            while right <= high:
                temp.append(nums[right])
                right += 1

            for i in range(len(temp)):
                nums[low + i] = temp[i]

        def mergeSort(low, high):
            if low >= high:
                return 0

            mid = (low + high) // 2

            count = 0

            count += mergeSort(low, mid)
            count += mergeSort(mid + 1, high)
            count += countPairs(low, mid, high)

            merge(low, mid, high)

            return count

        return mergeSort(0, len(nums) - 1)